[CORE-9704] Transactional control batch removal final final no more updates final v2

WillemKauf opened the pull request:
Removes transactional control batches during compaction (i.e makes them a "filterable" batch).

For a description of "filterable" and "compactible" batches, there was an old cover letter here that described these categories https://github.com/redpanda-data/redpanda/pull/26661:

> The key point here is that in the future, transactional control batches will be eligible for cleanup once their deletion window passes (`self_compact_timestamp + delete.retention.ms`). Previously, these batches were not considered "compactible"- meaning, we wouldn't even try to filter them through the `copy_data_segment_reducer`. Though these batches are still not "compactible" (i.e eligible for removal due to the existence of a record with the same key later in the log), they are removable by this deletion window- meaning they need to be considered as "filterable". To this effect, the logic of `should_keep()` and `copy_data_segment_reducer::filter()` is altered slightly to allow these batches into the code path that could potentially filter them out. "compactible" batches are still not considered by `should_keep::is_latest_key()`- they are either removed by deletion logic (this is still currently left unimplemented for transactional batches in this PR), or kept.

"Filterable" refers to all batches that can be removed by the compaction process, either by deduplication or by other means (e.g. expired tombstones or control batches).

"Compactible" refers to batches that contain records that can be removed due to the presence of a newer record (i.e a record with a higher offset) with the same key in the log. 

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

### Features

* Transactional control batches will now be removed by compaction of local `segment`s according to the time horizon set by `delete.retention.ms`. This feature can be disabled by setting the cluster config `log_compaction_disable_tx_batch_removal`.

WillemKauf committed:
`storage`: add `control_batches_discarded` to `compaction::stats`

To help track removed control batches during local storage compaction.

WillemKauf committed:
`storage`: revert addition of `max_removed_offset`

This was going to be used at one point for coordinated compaction
& safe tombstone removal, but not anymore.

Remove it from the tree.

WillemKauf committed:
`compaction`: rename `is_compactible_control_batch` to `is_removable...`

"Compactible" is now a term that suggests a record may be removed due to
de-duplication (i.e a record with the same key and a higher offset in the log
exists).

Rename this function to "is_removable_control_batch()", since it returns `true`
for control batches that can be immediately removed without any consideration of
the rules of de-duplication.

WillemKauf committed:
`storage`: add `probe::_control_batches_removed`

A metric that informs the user of how many control batches have been
removed by compaction due to expiration (i.e past the `delete.retention.ms`
horizon).

WillemKauf committed:
`rptest`: add `LogCompactionTxRemovalTest`

WillemKauf committed:
`config`: add `log_compaction_disable_tx_batch_removal`

A safety hatch mechanism to disable the removal of transactional
control batches during compaction.

WillemKauf committed:
`storage`: don't index uncompactible batches in `map`

This check is the same, but is expressed in a more compaction
oriented manner.

WillemKauf committed:
`storage`: consider `filterability` not `compactibility` in `segment` indexing

We want to force compaction for batches that contain filterable records
(all records potentially removable by compaction, whether via de-duplication
or expiration), not just compactible ones.

For example, a `segment` which only contained control batches would
return `false` for `has_compactible_records()`, which means we may skip
over compacting it, even though there is the potential for removing the
control batches within due to `delete.retention.ms` enforcement.

WillemKauf committed:
`storage`: remove transactional control batches during compaction

Similar to tombstone removal, this is a two step process:

1. During self compaction, all relevant transactional batches
have their transactional bit unset (analogous to the first pass in
tombstone removal in which all prior records with the same key as a
tombstone are removed) and the `tx_fence` control batch is removed.
2. The next time a control batch (a `raft_data` batch with the control
bit set) is encountered past the `delete.retention.ms` expiration horizon,
it can be removed (analogous to the second pass in tombstone removal,
though this horizon is set by the `self_compaction_timestamp`, not the
`clean_compact_timestamp`, for the differences in step 1.).

Of course, both tombstone and transactional control batch removal both
depend now on the `max_tombstone_remove_offset` set by coordinated compaction,
so the condition in step 2 also relies on the `segment`'s last/stable/dirty
offset being less than or equal to this offset.

Transactional control batch removal during compaction can be disabled via
the use of the cluster config `log_compaction_disable_tx_batch_removal` as
a safety hatch mechanism.

WillemKauf commented on src/v/compaction/types.cc:
```diff
@@ -38,11 +38,15 @@ std::ostream& operator<<(std::ostream& o, const stats& s) {
       s.records_discarded,
-      s.non_compactible_batches);
+      s.non_compactible_batches,
+      s.control_batches_discarded > 0
```
Conditional here because this `compaction` filter is also being used in cloud topics compaction, where `control_batches_discarded` will always == 0.

WillemKauf commented on src/v/storage/compaction_reducers.cc:
```diff
@@ -488,9 +488,7 @@ ss::future<ss::stop_iteration>
-    auto is_control = header.attrs.is_control();
-    auto type = header.type;
-    if (is_control || !compaction::is_filterable(type)) {
+    if (!compaction::is_compactible(header)) {
```
Same condition as `is_control || !compaction::is_filterable(type)`

WillemKauf commented on src/v/storage/segment_deduplication_utils.cc:
```diff
@@ -42,6 +43,13 @@ ss::future<ss::stop_iteration> put_entry(
     auto offset = idx_entry.offset + model::offset_delta(idx_entry.delta);
+    auto ctrl = is_control_from_enhanced_key(idx_entry.key);
+    auto bt = batch_type_from_enhanced_key(idx_entry.key);
+    if (ctrl || !compaction::is_filterable(bt)) {
```
Same condition as above (`is_control || !compaction::is_filterable(type)` evaluates to the same as `!compaction::is_compactible(header)`)

WillemKauf commented on src/v/storage/segment_utils.h:
```diff
@@ -360,44 +372,44 @@ ss::future<bool> should_keep(
         may_have_tombstone_records = true;
     }
 
-    if (b.contains_transactional_data() && keep) {
```
dead code path (control batches are either kept or removed at this point) and also the wrong check (we wanted to check `b.header().attrs.is_control()`)

WillemKauf commented on src/v/storage/segment_utils.h:
```diff
@@ -318,18 +313,19 @@ auto with_segment_reader_handle(segment_reader_handle handle, Func func) {
-//   2. Tombstone records past the removal horizon set by
+//   2. Expired tombstone records past the removal horizon set by
+//      `delete.retention.ms`
+//   2. Expired control batches past the removal horizon set by
```
ha yeah thats true

andrwng commented on src/v/storage/compaction_key.h:
```diff
@@ -21,4 +21,15 @@ namespace storage {
 compaction::compaction_key enhance_key(
   model::record_batch_type type, bool is_control_batch, bytes_view key);
 
+// Returns the `record_batch_type` from an enhanced `compaction_key`.
```
Are these changes more about improving key offset map efficiency, or are they relevant to control batch removal as well? Wonder if this can be pulled out of this PR

WillemKauf replied:
Yeah maybe this isn't worth adding.

andrwng commented on src/v/storage/segment_utils.h:
```diff
@@ -326,10 +326,9 @@ auto with_segment_reader_handle(segment_reader_handle handle, Func func) {
   bool past_tombstone_delete_horizon) {
     // Compactible control batches are always removable
-    if (compaction::is_compactible_control_batch(ntp, b.header().type)) {
+    if (compaction::is_compactible_control_batch(b.header().type)) {
```
Does this commit need to be reordered to be safe? I thought compactible control batches _aren't_ safe to be removable based on some horizon from Raft

WillemKauf replied:
This commit in isolation only affects how `tx_fence` batches are treated, which is still safe (since a consumer/broker seeing a `raft_data` batch with the transactional bit set will treat it as if they had seen a `tx_fence` batch, IIRC)

andrwng commented on src/v/storage/segment_utils.h:
```diff
@@ -326,10 +326,9 @@ auto with_segment_reader_handle(segment_reader_handle handle, Func func) {
 inline bool can_discard(
   const model::record_batch& b,
   const model::record& r,
-  const model::ntp& ntp,
```
nit: please update the above comment

WillemKauf replied:
Above comment is still correct.

andrwng commented on src/v/compaction/utils.cc:
```diff
@@ -18,17 +18,11 @@
-bool is_compactible_control_batch(
-  const model::ntp& ntp, const model::record_batch_type batch_type) {
+bool is_compactible_control_batch(const model::record_batch_type batch_type) {
     // Control batches in consumer offsets are special compared to
```
nit: please update this comment

WillemKauf replied:
The comment is still correct- batch types prefixed with `group_*_tx` are only found in the `__consumer_offsets` topic.

andrwng replied:
Ah makes sense, thanks!

andrwng commented on src/v/storage/segment_deduplication_utils.cc:
```diff
@@ -42,6 +43,13 @@ ss::future<ss::stop_iteration> put_entry(
   bool& fully_indexed) {
     auto offset = idx_entry.offset + model::offset_delta(idx_entry.delta);
+    auto ctrl = is_control_from_enhanced_key(idx_entry.key);
+    auto bt = batch_type_from_enhanced_key(idx_entry.key);
```
This seems to be doing a bunch of memcpy work on every key for something that feels rare. Is this important to do, or is it just an optimization to improve key offset map utilization

WillemKauf replied:
Yeah, agreed. Removing.

andrwng commented on src/v/storage/compaction_reducers.cc:
```diff
@@ -335,6 +335,7 @@ ss::future<ss::stop_iteration> copy_data_segment_reducer::filter_and_append(
     _acc += header_size;
     // do not set broker_timestamp in this index, leave the operation to the
     // caller who has more context
+    bool filterable_batch = compaction::is_filterable(batch.header().type);
```
>We want to force compaction for batches that contain filterable records
>(all records potentially removable by compaction, whether via de-duplication
>or expiration), not just compactible ones.

nit: can you please elaborate why this is the case?

WillemKauf replied:
Because otherwise, we may skip over compacting these segments in sliding window compaction or elsewhere where we check `segment->has_compactible_records()`

andrwng replied:
Thanks. Just making sure, is this understanding correct? :

There may be segments that contain only control batches, which are filterable but not compactible. These would previously return `has_compactible_records() = false`, and would cause us to skip compacting this segment. But "compaction" is really more "compact and remove" and we actually _do_ have removable control batches, so we need to indicate that to the index.

If it is, this level of walkthrough would be helpful to stick in the commit message, and it's probably also worth adding a brief comment about here because it may be confusing to future readers why we aren't counting compactible_records with is_compactible

andrwng replied:
Really I guess this is meant to convey that this batch has further compaction work to be done on it, in which case maybe that intent would be more clearly written `is_compactible || is_control` instead of `is_filterable`, though I don't feel strongly about it

WillemKauf replied:
Yes, your understanding is correct. I'll add a more descriptive comment to the commit message.

andrwng replied:
Thanks for confirming. Is it also correct in understanding that when we upgrade, we're not going to remove anything from segments that were written before 25.3 that contain only control batches, because their index will have `has_compactible_records = false`?

WillemKauf replied:
> we're not going to remove anything from segments that were written before 25.3 that contain only control batches, because their index will have has_compactible_records = false?

No, that will be incorrect, thanks to `may_have_transactional_batches` defaulting to `true` or containing `true` due to compaction detecting presence of existing control batches. I am writing an upgrade test as we speak but am heavily slowed by local testing issues.

https://redpandadata.slack.com/archives/C07HD9U0EHL/p1761078770322659

andrwng commented on src/v/compaction/utils.h:
```diff
@@ -17,7 +17,12 @@
+// consider either deduplication (compactible) or `delete.retention.ms` (a form
+// of filtering) to decide if these batches can be removed. They can be
+// indiscriminately removed when they are seen.
+bool is_removable_control_batch(const model::record_batch_type batch_type);
```
It's not really because this PR per se, but I'm finding it really difficult to follow the conceptual difference between filterable, compactible, and removable_control without going in and seeing their call sites. I'm wondering if there's a more intuitive way to describe the differences.

Maybe it'd help if we categorized the batch types / attributes, like:
- control batches (where the control bit is set)
- removable control batch (also a control batch, but specific to transactions?)
- STM batches
- data batches
- compaction placeholder batches

and described what capabilities each have, or don't have w.r.t compaction.

WillemKauf replied:
The code comments here in `utils.h` were meant to provide all the info required to grokking these differences- I can try to add another comment that comes from the opposite direction (describing batches and what category they are considered rather than describing the category and providing some examples of batches) if you feel that these comments in conjunction with the function definitions don't provide enough clarity as to how these batches are treated.

andrwng replied:
Yea I read them but still found it tough to build an intuition about when we'd want to use is_compactible vs is_filterable.

WillemKauf replied:
Yeah, I agree that it isn't super intuitive. I'm not sure if there is any better way to encode this categorization of batches but it may just be a result of the non-greenfield utilities here.

I'll think of a nicer way to clean this up in the future, if there is one.

andrwng commented on src/v/storage/compaction_reducers.cc:
```diff
@@ -201,6 +201,17 @@ copy_data_segment_reducer::filter(model::record_batch batch) {
+              "Removing transactional bit for raft batch {}",
+              header);
+            header.attrs.remove_transactional_type();
+            header.reset_size_checksum_metadata(batch.data());
```
Maybe a dumb q, but what's the significance of removing the transactional bit? What happens if we don't do this?

WillemKauf replied:
To prevent brokers from reading that this raft data is part of a transaction- making it indistinguishable from a regularly committed `raft_data` batch is important IIRC. cc: @bharathv for a more detailed explanation.

bharathv replied:
IIUC this is not really needed for correctness. First transactional data batch in a transaction marks the beginning of a transaction (this is true for Kafka clients to). Doing this here means we are shrinking the scope of the transaction and in most cases retaining the end control batch as the only batch remaining in transaction to be removed as a part of phase 2 of compaction.

On  disadvantage tho is we have to recompute the checksum which is extra CPU cycles than before.

WillemKauf replied:
> IIUC this is not really needed for correctness.

>On[e] disadvantage tho is we have to recompute the checksum which is extra CPU cycles than before.

So there's no upside and only one downside to removing the transactional bit & recomputing the checksum? If so, this could just be removed, but I seem to recall from prior conversations around this that the first pass should be removing the fence batch and unsetting this bit for correctness.

bharathv replied:
chatted with Willem offline, so we need this correctness actually.. imagine a simple case where we have a single transaction (one fence, one data batch and one commit), if the control batch removes takes it usual course and we don't unset the transactional bit, the final state will be a data batch with the tx bit set and that is effectively an open transaction..

andrwng replied:
Got it, so previously we could leave them as is because we always left behind the commit batch, and now that we're removing the commit/abort batches, we need to make sure the transaction status of the data batches doesn't depend on the commit/aborts.

And just thinking out loud, in the upgrade from 25.2, where we've self compacted in 25.2 (where the data batches are left with the transactional bit set) and then we upgrade.. it's safe to proceed with control batch removal because in the process of removing them we'll also remove any/all transaction bits as a part of the 25.3 compaction

WillemKauf replied:
Yes, your understanding is correct.

Bharath and I just discussed a potential issue that is now addressed in a new commit which changes `has_transactional_batches` which had a default value of `false` to `may_have_transactional_batches` with a default value of `true`. The commit message is copied here:

```
    `storage`: rename `has_transactional_batches` to `may_have...`
    
    For historical data, we need to consider this value `true` by default-
    consider a multi-segment transaction across segments [1] and [2],
    where [1] contains the `tx_fence`, data batches, and [2] contains the `commit`
    marker. Also assume that for whatever reason, [1] has not been recompacted
    since this `has_transactional_batches` marker was added, but [2] has,
    meaning [1] has `has_transactional_batches=false` and [2] has it as `true`.
    
    When we compact the log, [2] is compacted and potentially has its `commit`
    marker removed while the `tx_fence` in [1] is left unremoved, and the
    `raft_data` batches in [1] are _also_ left with the transactional bit set.
    
    We could get into a situation in which the `raft_data` batches are read,
    and then some `abort` marker in a segment later in the log is mistakenly
    applied, since the `commit` marker is now missing.
    
    Prevent this situation by ensuring that both segments [1] and [2] would
    have `may_have_transactional_batches=true` set even if [1] has not been
    recompacted since the addition of this marker, leading to compaction of [1]
    BEFORE compaction of [2].
```

andrwng replied:
Nice catch.

    When we compact the log, [2] is compacted and potentially has its `commit`
    marker removed while the `tx_fence` in [1] is left unremoved, and the
    `raft_data` batches in [1] are _also_ left with the transactional bit set.

Just for my own understanding, why isn't the removal of the `commit` marker prevented by coordinated compaction if [1] hasn't been cleanly compacted yet? Or is the problem that the accounting of "cleanly compacted" from 25.2 segments not accounting for transactional batches?

WillemKauf replied:
Yeah, there is no concept of "cleanly compacted" here, we go off of the fact that segments are self compacted from earliest to newest and rely on the `self_compact_timestamp + delete.retention.ms` for the removal of control batches.

andrwng reviewed (commented):
I found this a bit tough to review without a refresher in transaction batches and their current handling. Could you please add a cover letter giving some context about the changes to the categorization of batch types, and why that's required for tx batch removal?

WillemKauf commented:
Force push to:
* Remove `_from_enhanced_key` functions and usage

andrwng commented on src/v/compaction/utils.cc:
Idea: `bool type_has_key()`

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74848-->
<!--ba8f6149-ec77-4159-b507-2bf79e8c52b0-->
<!--61683032-5a5d-4255-81c7-6fc8b68571ef-->
<!--1da9cc52-6ae8-4f43-a01f-c71890d43d90-->
<!--ce9798f5-eee1-4708-a4d8-6b732f220842-->
<!--title-->
## Retry command for [Build#74848](https://buildkite.com/redpanda/redpanda/builds/74848)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":false}
```
<!--/body-->

WillemKauf commented:
/ci-repeat 1
skip-redpanda-build
skip-units
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":false}

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74876-->
<!--03047a5f-727a-4439-8898-80a24fea247d-->
<!--39c68145-366b-4c3b-924c-1ffa737c3c51-->
<!--title-->
## Retry command for [Build#74876](https://buildkite.com/redpanda/redpanda/builds/74876)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":false}
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--d5106ad7-0e96-413b-b273-de8abfc273b4-->
<!--TEST RESULTS-->
<!--4a7fac29-91f1-4e19-be09-a315ec71e344-->
<!--TEST RESULTS-->
<!--60e880a4-a250-4a34-ab5f-febb4ee8fa11-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74876">build#74876</a></summary>


| test_class                 | test_method          | test_arguments      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 | test_history                                                                                                                                                  |
|:---------------------------|:---------------------|:--------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74876#019a1396-2df5-4b87-a74f-e24a353d7657 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74876#019a1396-2df9-4282-8d1d-4f8fb75c665c | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/74876#019a1396-2df5-4b87-a74f-e24a353d7657 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/74876#019a1396-2df9-4282-8d1d-4f8fb75c665c | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74917">build#74917</a></summary>


| test_class                      | test_method         | test_arguments                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                 |
|:--------------------------------|:--------------------|:--------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Datalake3rdPartyMaintenanceTest | test_e2e_basic      | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1, "query_engine": "spark"} | integration | https://buildkite.com/redpanda/redpanda/builds/74917#019a16af-bc2d-44dd-9fc6-a698ec30f99b | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=Datalake3rdPartyMaintenanceTest&test_method=test_e2e_basic |
| MountUnmountIcebergTest         | test_simple_remount | {"cloud_storage_type": 1}                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/74917#019a16b8-1567-4b3b-820a-8712d1364262 | FLAKY         | 17/21    | upstream reliability is '91.98717948717949'. current run reliability is '80.95238095238095'. drift is 11.0348 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount    |
| PartitionMovementTest           | test_empty          | {"num_to_upgrade": 0}                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74917#019a16af-bc36-4f9e-baa1-6fa2a45eb245 | FLAKY         | 18/21    | upstream reliability is '94.09937888198758'. current run reliability is '85.71428571428571'. drift is 8.38509 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_empty               |
| PartitionMovementTest           | test_empty          | {"num_to_upgrade": 0}                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74917#019a16b8-156a-4d6f-89d4-552943181a48 | FLAKY         | 17/21    | upstream reliability is '94.09937888198758'. current run reliability is '80.95238095238095'. drift is 13.147 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_empty               |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74987">build#74987</a></summary>


| test_class                                        | test_method                                       | test_arguments                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                           |
|:--------------------------------------------------|:--------------------------------------------------|:---------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests                              | test_rapid_shadow_link_toggling                   | null                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74987#019a18d7-49c7-48de-8e04-44a11cf7584c | FLAKY         | 16/21    | upstream reliability is '90.86251067463706'. current run reliability is '76.19047619047619'. drift is 14.67203 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_rapid_shadow_link_toggling     |
| ShadowLinkBasicTests                              | test_topic_properties_replicated                  | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74987#019a18d7-49c3-408d-b464-bf7dae619aff | FLAKY         | 19/21    | upstream reliability is '94.92753623188406'. current run reliability is '90.47619047619048'. drift is 4.45135 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_topic_properties_replicated    |
| RedpandaNodeOperationsSmokeTest                   | test_node_ops_smoke_test                          | {"cloud_storage_type": 1, "mixed_versions": true}                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74987#019a18d7-49c8-4f1b-b42c-dc263184969c | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
| src/v/raft/tests/group_configuration_test         | src/v/raft/tests/group_configuration_test         |                                                                                                                | unit        | https://buildkite.com/redpanda/redpanda/builds/74987#019a1872-3fdf-4e4a-aea2-36e8d55d5287 | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                        |
| src/v/raft/tests/raft_replicate_rpbench_test      | src/v/raft/tests/raft_replicate_rpbench_test      |                                                                                                                | unit        | https://buildkite.com/redpanda/redpanda/builds/74987#019a1872-3fdf-4e4a-aea2-36e8d55d5287 | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                        |
| src/v/storage/mvlog/tests/entry_stream_utils_test | src/v/storage/mvlog/tests/entry_stream_utils_test |                                                                                                                | unit        | https://buildkite.com/redpanda/redpanda/builds/74987#019a1872-3fdf-4e4a-aea2-36e8d55d5287 | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                        |
| src/v/storage/mvlog/tests/file_test               | src/v/storage/mvlog/tests/file_test               |                                                                                                                | unit        | https://buildkite.com/redpanda/redpanda/builds/74987#019a1872-3fdf-4e4a-aea2-36e8d55d5287 | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                        |
| src/v/storage/tests/segment_concatenation_test    | src/v/storage/tests/segment_concatenation_test    |                                                                                                                | unit        | https://buildkite.com/redpanda/redpanda/builds/74987#019a1872-3fdf-4e4a-aea2-36e8d55d5287 | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                        |
| storage_test_fixture                              | max_compaction_lag                                |                                                                                                                | unit        | https://buildkite.com/redpanda/redpanda/builds/74987#019a1872-3fdf-4e4a-aea2-36e8d55d5287 | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                        |
| storage_test_fixture                              | test_offset_range_size2_compacted                 |                                                                                                                | unit        | https://buildkite.com/redpanda/redpanda/builds/74987#019a1872-3fdf-4e4a-aea2-36e8d55d5287 | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                        |
</details>
<!--/body-->

WillemKauf commented:
PR is green, `ShadowLinkingRandomOpsTest` is red.

bashtanov commented on src/v/storage/segment_utils.cc:
```diff
@@ -1388,7 +1395,15 @@ ss::future<bool> mark_segment_as_finished_self_compaction(
 bool is_past_transaction_batch_delete_horizon(
   ss::lw_shared_ptr<segment> seg, const compaction::compaction_config& cfg) {
-    if (seg->has_self_compact_timestamp() && cfg.tx_retention_ms.has_value()) {
+    if (config::shard_local_cfg().log_compaction_disable_tx_batch_removal()) {
```
This config parameter is introduced in a further commit, so I'm not sure if the current commit will compile. Maybe reorder the commits?

There's a handy script to run a check against each commit in the PR, e.g. to check intermediate commits are buildable: `vtools/scripts/crawl-down.sh bazel build //...`

WillemKauf replied:
Oops, not sure how these got re-ordered, thanks.

WillemKauf commented:
Force push to correct commit history

bharathv commented on src/v/compaction/utils.cc:
```diff
@@ -18,17 +18,11 @@
-             batch_type == model::record_batch_type::tx_fence
-             && model::is_consumer_offsets_topic(ntp))
+    // Fence batches can also be safely removed.
+    return batch_type == model::record_batch_type::tx_fence
```
I was wondering if this should be predicated on the configuration "log_compaction_disable_tx_batch_removal" introduced in the later comments. If set to true, we should retain these batches, no?

WillemKauf replied:
Sure, if we wanted `log_compaction_disable_tx_batch_removal` to gate the new behavior completely we would add a check here as well. Though unless we leave `ntp` in this path, we will have a difference in behavior in terms of not removing `tx_fence` batches in `__consumer_offsets` when `log_compaction_disable_tx_batch_removal=true`.

WillemKauf replied:
LMK what you would prefer (check or no check)

bharathv replied:
I'd prefer to check so we can fallback to original behavior fully incase of any emergencies

WillemKauf replied:
Done

bharathv commented on src/v/storage/compaction_reducers.cc:
```diff
@@ -505,8 +503,9 @@ map_building_reducer::operator()(model::record_batch batch) {
-       type,
-       is_control](const model::record& r) -> ss::future<ss::stop_iteration> {
+       type = header.type,
+       is_control = header.attrs.is_control()](
```
just an observation, is_control is always false, right?

WillemKauf replied:
Yes, that's true. I'll remove this and just add an explicit `false` to this path if you think that would be better for readability.

andrwng commented on src/v/storage/segment_utils.h:
```diff
@@ -331,6 +334,11 @@ inline bool can_discard(
     }
 
+    // Deal with transactional control batch removal
+    if (b.header().attrs.is_control() && past_tx_delete_horizon) {
```
Another dumb q, are there non-transaction control batches?

WillemKauf replied:
> non-transaction control batches

No, `is_control()` is only set for transactional batches, whether that is a `tx_fence` batch, a `raft_data` batch with `abort/commit` bits set, or `group_tx_*` batch types (and now, I guess also placeholder batches for cloud topics, but those aren't relevant here)

bharathv commented on tests/rptest/tests/log_compaction_test.py:
```diff
@@ -628,3 +630,153 @@ def seen_compacted_segments():
+            if record_or_batch["type_name"] == "tx_fence":
+                num_fence_batches += 1
+
+        assert num_control_batches == 0, (
```
I'm wondering how this is possible with the default delete retention ms (the test doesn't seem to override this in rp_conf).

WillemKauf replied:
`LogCompactionTestBase` sets `delete_retention_ms = 3000` on L57.

bharathv reviewed (commented):
It'd be interesting to see the results of chaos tests (especially since we are removing fence batches and transactional bits in first pass).

@WillemKauf a general question. Say we have transactions only use cases (no tombstones), the compaction is more CPU/IO intensive with these changes right? As in we do a 2 pass compaction (vs 1 before), so more IO and also we constantly rechecksum the batches to undo the transactional bit, so overall its more heavier in terms of both CPU/IO post upgrade, right?

WillemKauf commented:
> It'd be interesting to see the results of chaos tests (especially since we are removing fence batches and transactional bits in first pass).

Are these not ran by default in CI? I can try them locally.

WillemKauf commented:
> Say we have transactions only use cases (no tombstones), the compaction is more CPU/IO intensive with these changes right? As in we do a 2 pass compaction (vs 1 before), so more IO and also we constantly rechecksum the batches to undo the transactional bit, so overall its more heavier in terms of both CPU/IO post upgrade, right?

Compaction might be slightly more intensive in that we will recompact a segment that is already compacted if we now believe we can remove transactional batches. But this is the same case as say, a segment with tombstones.

andrwng reviewed (commented):
Code LGTM but will leave final signoff to Alexey or Bharath

bharathv commented:
> > It'd be interesting to see the results of chaos tests (especially since we are removing fence batches and transactional bits in first pass).
> 
> Are these not ran by default in CI? I can try them locally.

I don't think so, it has to be manual (crafting the correct set of buildkite CI options) if you wan to run against unmerged changes. may be ok to run it post merge.

WillemKauf commented:
> I don't think so, it has to be manual (crafting the correct set of buildkite CI options) if you wan to run against unmerged changes. may be ok to run it post merge.

The "transactions" suite that used to be part of the chaos testing AFAIK is ran in CI, and all of those tests have passed.

Unless you are indeed referring to the remaining `chaos_test` suite (which is not run in CI under the `test_suite_quick.yml` suite)

bharathv commented:
> > I don't think so, it has to be manual (crafting the correct set of buildkite CI options) if you wan to run against unmerged changes. may be ok to run it post merge.
> 
> The "transactions" suite that used to be part of the chaos testing AFAIK is ran in CI, and all of those tests have passed.
> 
> Unless you are indeed referring to the remaining `chaos_test` suite (which is not run in CI under the `test_suite_quick.yml` suite)

right I meant the full chaos_test suite (it has some tests with compaction and partition movements in a tight loop).

WillemKauf commented:
/ci-repeat
skip-redpanda-build
skip-units
rp_repo=ci
tests/rptest/test_suite_chaos.yml

WillemKauf commented:
Force push to:
* Add requested code comments
* Gate removal of `tx_fence` behind `log_compaction_disable_tx_batch_removal`

WillemKauf committed:
`storage`: rename `has_transactional_batches` to `may_have...`

For historical data, we need to consider this value `true` by default-
consider a multi-segment transaction across segments [1] and [2],
where [1] contains the `tx_fence`, data batches, and [2] contains the `commit`
marker. Also assume that for whatever reason, [1] has not been recompacted
since this `has_transactional_batches` marker was added, but [2] has,
meaning [1] has `has_transactional_batches=false` and [2] has it as `true`.

When we compact the log, [2] is compacted and potentially has its `commit`
marker removed while the `tx_fence` in [1] is left unremoved, and the
`raft_data` batches in [1] are _also_ left with the transactional bit set.

We could get into a situation in which the `raft_data` batches are read,
and then some `abort` marker in a segment later in the log is mistakenly
applied, since the `commit` marker is now missing.

Prevent this situation by ensuring that both segments [1] and [2] would
have `may_have_transactional_batches=true` set even if [1] has not been
recompacted since the addition of this marker, leading to compaction of [1]
BEFORE compaction of [2].

andrwng commented:
> > > I don't think so, it has to be manual (crafting the correct set of buildkite CI options) if you wan to run against unmerged changes. may be ok to run it post merge.
> > 
> > 
> > The "transactions" suite that used to be part of the chaos testing AFAIK is ran in CI, and all of those tests have passed.
> > Unless you are indeed referring to the remaining `chaos_test` suite (which is not run in CI under the `test_suite_quick.yml` suite)
> 
> right I meant the full chaos_test suite (it has some tests with compaction and partition movements in a tight loop).

Yeah, also probably makes sense to make `delete.retention.ms` some reasonably low value that would actually exercise tx marker removal in those tests
